### PR TITLE
extract all URLs from a text

### PR DESCRIPTION
**Pull request by awhitesong**
Here is the basic implementation for this issue. I have created a new function `lua_url_all`. Please review if any changes. Thanks.

### CODE REVIEW

File: src/lua/lua_url.c [hunk 1]
```diff
@@ -370,6 +372,49 @@ lua_url_create (lua_State *L)
 }
 
 static gint
+lua_url_all (lua_State *L)
+{
+	struct rspamd_url *url;
+	rspamd_mempool_t *pool = rspamd_lua_check_mempool (L, 1);
+	const gchar *text,*end;
+	gint length,i;
```
Comment (vstakhov) [inline src/lua/lua_url.c]: `i` used uninitialized.

File: src/lua/lua_url.c [hunk 2]
```diff
@@ -370,6 +372,49 @@ lua_url_create (lua_State *L)
...
+	const gchar **pos;
+
+	if (pool == NULL) {
+		lua_pushnil (L);
+	}
+	else {
+		text = luaL_checkstring (L, 2);
+
+		if (text != NULL) {
+			length = strlen(text);
```
Comment (vstakhov) [inline src/lua/lua_url.c]: You can get length of the string by just using `luaL_checklstring (L, 2, &length);`

File: src/lua/lua_url.c (hunk 1 above)
Comment (vstakhov) [inline src/lua/lua_url.c]: length should have type `size_t`.

File: src/lua/lua_url.c [hunk 3]
```diff
@@ -370,6 +372,49 @@ lua_url_create (lua_State *L)
...
+			*pos=text;
+			end=text+length;
+			lua_newtable(L);
+			while(*pos<=end){
```
Comment (vstakhov) [inline src/lua/lua_url.c]: some style comment: there should be space before opening bracket and spaces before and after `<=`

File: src/lua/lua_url.c [hunk 4]
```diff
@@ -370,6 +372,49 @@ lua_url_create (lua_State *L)
...
+				url = rspamd_url_get_next (pool, text, pos, NULL);				
+
+				if (url!=NULL) {
```
Comment (vstakhov) [inline src/lua/lua_url.c]: But if your `url` is NULL, you'll get a gap in lua table indicies, e.g. [1, 2, 4, 6]

File: src/lua/lua_url.c [hunk 5]
```diff
@@ -370,6 +372,49 @@ lua_url_create (lua_State *L)
...
+					lua_pushinteger (L, i);
+					lua_pushlstring (L, url->string, url->urllen);
+					lua_settable (L, -3);
```
Comment (vstakhov) [inline src/lua/lua_url.c]: If `url` is NULL you should just call `break` there.

File: src/lua/lua_url.c [hunk 6]
```diff
@@ -370,6 +372,49 @@ lua_url_create (lua_State *L)
...
+			length = strlen(text);
+			*pos=text;
+			end=text+length;
+			lua_newtable(L);
+			while(*pos<=end){
+				url = rspamd_url_get_next (pool, text, pos, NULL);				
+
+				if (url!=NULL) {
+					lua_pushinteger (L, i);
+					lua_pushlstring (L, url->string, url->urllen);
```
Comment (vstakhov) [inline src/lua/lua_url.c]: Shouldn't you push `rspamd{url}` here? See https://github.com/vstakhov/rspamd/blob/master/src/lua/lua_url.c#L358

**Comment by awhitesong** [comment]
Hi, so i have modified and made specified changes. Just a little documentation is left for now. Please tell if any further changes needed. Thanks.

File: src/lua/lua_url.c [hunk 7]
```diff
@@ -370,6 +372,55 @@ lua_url_create (lua_State *L)
...
+
+				if (url != NULL) {
+					lua_url = lua_newuserdata (L, sizeof (struct rspamd_lua_url));
+					rspamd_lua_setclass (L, "rspamd{url}", -1);
+					lua_url->url = url;
+					lua_pushinteger (L, i++);
+					lua_pushlstring (L, url->string, url->urllen);
+					lua_settable (L, -3);
+				}
+				else
```
Comment (vstakhov) [inline src/lua/lua_url.c]: Small style comment: in rspamd, every condition and loop *must* be enclosed within braces, even if it is one line. E.g.
```C
if (some)
   do();
else
   done();
```

should be written as:
```C
if (some) {
   do();
}
else {
   done();
}
```

**Comment by awhitesong** [comment]
Changed the `if else` formatting.

**Comment by vstakhov** [comment]
According to http://pgl.yoyo.org/luai/i/lua_settable `lua_settable` pops both key and value from the stack. I doubt that extra `lua_pop` is required in this case.

**Pull request merged**